utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+#ifndef ROCKSDB_LITE
+
+#include "utilities/blob_db/blob_db_impl.h"
+#include <algorithm>
+#include <chrono>
+#include <cinttypes>
+#include <ctime>
+#include <iomanip>
+#include <limits>
+#include <memory>
+
+#include "db/db_impl.h"
+#include "db/filename.h"
+#include "db/write_batch_internal.h"
+#include "rocksdb/convenience.h"
+#include "rocksdb/env.h"
+#include "rocksdb/iterator.h"
+#include "rocksdb/utilities/stackable_db.h"
+#include "rocksdb/utilities/transaction.h"
+#include "table/block.h"
+#include "table/block_based_table_builder.h"
+#include "table/block_builder.h"
+#include "table/meta_blocks.h"
+#include "util/crc32c.h"
+#include "util/file_reader_writer.h"
+#include "util/instrumented_mutex.h"
+#include "util/timer_queue.h"
+#include "utilities/transactions/optimistic_transaction_db_impl.h"
+#include "utilities/transactions/optimistic_transaction_impl.h"
+
+namespace {
+
+void extendTTL(rocksdb::blobstorage::ttlrange_t* ttl_range, uint32_t ttl) {
+  ttl_range->first = std::min(ttl_range->first, ttl);
+  ttl_range->second = std::max(ttl_range->second, ttl);
+}
+
+void extendTimestamps(rocksdb::blobstorage::tsrange_t* ts_range, uint64_t ts) {
+  ts_range->first = std::min(ts_range->first, ts);
+  ts_range->second = std::max(ts_range->second, ts);
+}
+
+void extendSN(rocksdb::blobstorage::snrange_t* sn_range,
+              rocksdb::SequenceNumber sn) {
+  sn_range->first = std::min(sn_range->first, sn);
+  sn_range->second = std::max(sn_range->second, sn);
+}
+}  // end namespace
+
+namespace rocksdb {
+
+namespace blobstorage {
+Random blob_rgen(static_cast<uint32_t>(time(nullptr)));
+
+void BlobDBFlushBeginListener::OnFlushBegin(DB* db, const FlushJobInfo& info) {
+  if (impl_) impl_->OnFlushBeginHandler(db, info);
+}
+
+WalFilter::WalProcessingOption BlobReconcileWalFilter::LogRecordFound(
+    unsigned long long log_number, const std::string& log_file_name,
+    const WriteBatch& batch, WriteBatch* new_batch, bool* batch_changed) {
+  return WalFilter::WalProcessingOption::kContinueProcessing;
+}
+
+bool blobf_compare_ttl::operator()(const std::shared_ptr<BlobFile>& lhs,
+                                   const std::shared_ptr<BlobFile>& rhs) const {
+  if (lhs->ttl_range_.first < rhs->ttl_range_.first) return true;
+
+  if (lhs->ttl_range_.first > rhs->ttl_range_.first) return false;
+
+  return lhs->BlobFileNumber() > rhs->BlobFileNumber();
+}
+
+void EvictAllVersionsFilter::Callback(int level, const Slice& key,
+                                      ValueType value_type,
+                                      const Slice& existing_value,
+                                      const SequenceNumber& sn,
+                                      bool is_new) const {
```

Comment:
I thought the solution we decided to take was not to rely on compaction filter but to directly use internal iterators. What problem did you see using that solution?